### PR TITLE
Variable band

### LINKED ISSUE #324: Implement band for previous runs instead of dots for line chart

### PR DESCRIPTION
**Pull request by mj3cheun**
Adds band to variable panel and fixes incorrect aggregate function (mean avg value calculated incorrect).

Note: when testing keep in mind that the bands are only enabled when there are more than 5 runs

Fixes #324

Screen Shot 2021-08-04 at 5 17 18 PM

### CODE REVIEW

File: client/src/components/widgets/charts/MultiLinePlot.vue [hunk 1]
```diff
@@ -160,6 +161,17 @@
...
     }
 
+    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+    drawPolygon (svg, data): void {
+      svg.selectAll('polygon')
+        .data([data])
+        .enter().append('polygon')
+          .attr('points', d =>
+              d.map(d => [this.xScale(d.x), this.yScale(d.y)].join(',')).join(' '))
+          .attr('fill', '#647EA9')
```
Comment (RosaRomeroGomez) [inline client/src/components/widgets/charts/MultiLinePlot.vue:171]: Could we move this color to the `DEFAULT_STYLE` object?

File: client/src/views/Simulation/components/VariablesPanel.vue [hunk 2]
```diff
@@ -114,10 +115,10 @@
 
   const AGGREGATE_STYLE = {
     node: {
-      fill: Colors.NODES.AGGREGATE,
+      fill: '#EBCB8B',
```
Comment (RosaRomeroGomez) [inline client/src/views/Simulation/components/VariablesPanel.vue:118]: Why this change?

File: client/src/views/Simulation/components/VariablesPanel.vue [hunk 3]
```diff
@@ -164,20 +159,40 @@
...
       variables.map(variable => {
         variable.styles = variable.styles || [];
-        for (let i = 0; i < variable.values.length; i++) {
-          if (variable.values.length > 5) {
-            i !== variable.values.length - 1
-              ? variable.styles.push(mergeStyle(OTHER_STYLE, NO_LINE_STYLE))
-              : variable.styles.push(mergeStyle());
-          } else {
-            variable.styles.push(mergeStyle(i !== variable.values.length - 1 && OTHER_STYLE));
+        if (variable.values.length > 5) {
```
Comment (RosaRomeroGomez) [inline client/src/views/Simulation/components/VariablesPanel.vue:188]: I think we should have this arbitrary number of runs as a constant in this panel. Something like `THRESHOLD = 5`

File: client/src/store/modules/simulation.ts [hunk 4]
```diff
@@ -246,7 +246,7 @@ const mutations: MutationTree<HMI.SimulationState> = {
       // A variable.values are a list of runs each containing
       // a list of {x: step, y: value} for each step.
       const valuesOfAllRunsPerStep = {};
-      for (let run = 0; run < state.numberOfSavedRuns; run++) {
+      for (let run = 0; run < variable.values.length; run++) {
```
Comment (YohannParis) [inline client/src/store/modules/simulation.ts:249]: That would explain the wrong value. Cheers!

File: client/src/views/Simulation/components/VariablesPanel.vue [hunk 5]
```diff
@@ -164,20 +159,40 @@
     @Action hideAllVariables;
     @Action showAllVariables;
 
+    calcPolygon (variable: HMI.SimulationVariable): void {
```
Comment (YohannParis) [inline client/src/views/Simulation/components/VariablesPanel.vue:162]: I understand that for d3 this data is a Polygon, but for us this is a band of values.
Maybe I'm picky but it would be clearer if the function and variables were named after what it represent.
Unless you thought of another situation were we can use polygons?

File: client/src/views/Simulation/components/VariablesPanel.vue [hunk 6]
```diff
@@ -164,20 +159,40 @@
...
+      const numRuns = variable.values.length;
```
Comment (YohannParis) [inline client/src/views/Simulation/components/VariablesPanel.vue:163]: You only use `numRuns - 1`, so why not just set it as `const numRuns = variable.values.length - 1;`?

File: client/src/views/Simulation/components/VariablesPanel.vue [hunk 7]
```diff
@@ -164,20 +159,40 @@
...
+        run.forEach((point, pointIndex) => {
+          xArr[pointIndex] = point.x;
+          yMinArr[pointIndex] = Math.min(yMinArr[pointIndex] ?? point.y, point.y);
+          yMaxArr[pointIndex] = Math.max(yMaxArr[pointIndex] ?? point.y, point.y);
+        });
+      }
+
+      variable.polygon = [
+        ...yMinArr.map((y, i) => ({ x: xArr[i], y })),
+        ...yMaxArr.map((y, i) => ({ x: xArr[i], y })).reverse(),
```
Comment (YohannParis) [inline client/src/views/Simulation/components/VariablesPanel.vue:179]: Could you add a small comment explaining why you use `reverse()`, it's not obvious right away.

**Review by YohannParis — COMMENTED** [review]
Looks great, and it's really neat to see the band.

File: client/src/views/Simulation/components/VariablesPanel.vue (hunk 2 above)
Comment (mj3cheun) [reply client/src/views/Simulation/components/VariablesPanel.vue:118]: `Colors.NODES.AGGREGATE` has been deleted at some point so the colour code is no longer there causing a problem where the aggregate run looks the same as the current run

File: client/src/views/Simulation/components/VariablesPanel.vue (hunk 5 above)
Comment (mj3cheun) [reply client/src/views/Simulation/components/VariablesPanel.vue:162]: I would prefer to keep the word `polygon` there in some way because it is important in understanding why the data structure is structured the way it is. I see your point though. What do you think of using the fn name `calcBandPolygon`?

**Comment by YohannParis** [comment]
Thank you for al your changes. Especially that comment! It is above and beyond.

**Review by RosaRomeroGomez — APPROVED** [review]
Looks good, thanks!
